invenio_previewer/extensions/audio_videojs.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2023 Northwestern University.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Render audio player using videojs."""
+
+from flask import current_app, render_template
+
+from ..proxies import current_previewer
+from ..utils import dotted_exts
+
+# The audio file world is complex (but less so than video one), because support
+# depends on hardware, OS, software and browser. A reasonable choice of
+# supported extensions
+# (https://en.wikipedia.org/wiki/HTML5_audio#Supported_audio_coding_formats)
+# that are least likely lead to an unsupported warning for end-users are mp3,
+# wav, aac and flac, but this could be made changeable per instance.
+previewable_extensions = current_app.config.get(
+    "PREVIEWER_AUDIO_EXTS", ["mp3", "wav", "aac", "flac"]
+)
+
+
+def can_preview(file):
+    """Determine if the given file can be previewed."""
+    return file.has_extensions(*dotted_exts(previewable_extensions))
+
+
+def preview(file):
+    """Render the appropriate template with embed flag."""
+    data_setup = {
+        "controls": True,
+        "preload": "metadata",
+        "fill": True,
+        "audioOnlyMode": True,
+        **current_app.config.get("PREVIEWER_AUDIO_DATA_SETUP", {}),
+    }
+
+    return render_template(
+        "invenio_previewer/audio_videojs.html",
+        file=file,
+        data_setup=data_setup,
+        js_bundles=current_previewer.js_bundles + ["videojs_js.js"],
+        css_bundles=current_previewer.css_bundles + ["audio_videojs_css.css"],
+    )
```

Comment:
Both files `audio_videojs.py` and `video_videojs.py` have similar structures and functionalities, consider abstracting common logic to a shared utility or base class to follow DRY (Don't Repeat Yourself) principles.

Reply:
All previewers share the `previewable_extensions`, `can_preview`, `preview` structure, so that's normal. For audio/video, the previewers use different extensions, different config, different css file. There isn't much repetition that is not warranted; even with a common abstraction, boilerplate would be repeated like it is here, so I don't really see a better common abstraction.

... and even if there was I would probably not go for it since we only have 2 cases and we are not [repeating knowledge](https://media.pragprog.com/titles/tpp20/dry.pdf), haha! I am of the "3 or 4- strike, refactor" school of thought of [Martin Fowler and others](https://www.refactoring.com/). Creating an abstraction makes most sense when it is clear what the concept that is being abstracted is - and that usually reveals itself after more than 2 occurrences. DRY triggers me a bit :smile_cat: : we have to be very careful with it because it is a trade-off. It is less code repetition (after a threshold) and fragmentation in exchange for more coupling and abstraction/complexity.

Reply:
While I acknowledge your approach, early abstraction in audio_videojs.py and video_videojs.py might streamline future updates and ensure consistency. Strict adherence to the "3 or 4-strike" rule could overlook early abstraction benefits. Curious to hear others' thoughts on this balance.

Reply:
I don't think the abstraction is worth the effort.